### DISCUSSION TITLE
Site to Site Tunnel /libexec/ipsec/_updown: not found Error

### DISCUSSION THREAD

**Discussion by rfraser578** [gitbls/pistrong#17] [General]
Hello,
I added a Site to Site tunnel to an existing installation of Pistrong on a VPN Server which has been running for a couple of years for Client/Server type device connections. That system is running Raspbian GNU/Linux 10 (buster). Looking at the Server syslog, the tunnel seems to be created successfully, but then after 5 seconds it closes.

The initiating Site is running on a clean install of Raspbian GNU/Linux 11 (bullseye) and Pistrong. StrongSwan was installed using Pistrong and the apt method was chosen. The initiating Site shows the /libexec/ipsec/_updown: not found Error in the syslog. Please see below.

Dec  6 12:38:46 pivpnclient charon-systemd[301]: IKE_SA pivpn2-Tunnel[1] established between 192.168.1.189[pivpnclient.SE1-AB10.com]...xxx.xxx.xxx.xxx[pivpn2.SE1-AB10.com]
Dec  6 12:38:46 pivpnclient charon-systemd[301]: selected proposal: ESP:AES_GCM_16_128/NO_EXT_SEQ
Dec  6 12:38:47 pivpnclient charon-systemd[301]: CHILD_SA pivpn2-Tunnel{1} established with SPIs c8e9de9c_i c491d511_o and TS 192.168.29.0/24 === 
192.168.19.0/24
Dec  6 12:38:47 pivpnclient charon-systemd[301]: updown: /bin/sh: 1: /libexec/ipsec/_updown: not found
Dec  6 12:38:50 pivpnclient charon-systemd[301]: vici terminate IKE_SA 'pivpn2-Tunnel'
Dec  6 12:38:50 pivpnclient charon-systemd[301]: deleting IKE_SA pivpn2-Tunnel[1] between 192.168.1.189[pivpnclient.SE1-AB10.com]...xxx.xxx.xxx.xxx[
pivpn2.SE1-AB10.com]
Dec  6 12:38:50 pivpnclient charon-systemd[301]: IKE_SA deleted
Dec  6 12:38:50 pivpnclient charon-systemd[301]: updown: /bin/sh: 1: /libexec/ipsec/_updown: not found

I didn't notice any errors when installing or configuring the system for the new tunnel. Should both ends of the tunnel be running the same OS and versions of strongSwan?

Any advice or help would be appreciated.

**Comment by gitbls** [comment]
Sorry you're having an issue. I believe there's a buglet having to do with different OS at each end of the tunnel, but you shouldn't have to fix that (run the same OS)...unless there's a strongSwan incompatibility issue between the strongSwan in each OS distro (seems like a remote possibility to me).

That said, on to your problem. 

On the system where _updown isn't found, here's what to do (all on the Bullseye system):
```
sudo ls -l /usr/lib/ipsec/_updown
```
If that file exists, you can `sudoedit` or `sudo <your editor> /etc/swanctl/conf.d/Tunnel-<host1>-<host2>.conf` and locate the line starting with `updown` and replace "/libexec/ipsec/_updown" with "/usr/lib/ipsec/_updown". 

Save the file, then `sudo systemctl restart strongswan` *or* `sudo pistrong service restart` to restart strongSwan, and then try the connection again.

I'm pretty sure that on Buster (Debian 10) pistrong had to install strongSwan using a strongSwan tarball to get a version that by default works with systemd. The version in apt was not new enough to have been built with that default (change happened in 5.8). 

This appears to be confirmed by updown being set as it is on your system.

The strongSwan version in Bullseye was later than 5.8, and hence had the necessary systemd support, so no need to build from a tarball.

pistrong implicitly assumes that both ends are running the same version in that the _updown script will be in the same place on both systems which is clearly not the case in your scenario.

If I've guessed wrong, LMK and we'll do more diagnosis.

**Reply by rfraser578** [reply]
Hi Benn,
I edited the file per your instructions and I no longer get the file not found error. However, the tunnel is still closed down seemingly by request from the client (Bullseye) system.

I can do more diagnosis if you think it would be useful for knowledge and I've attached the two syslog extracts.

I think I should really just upgrade the old server running Buster and see where it goes from there. It might be a better solution.

Kind regards
[Server side syslog after fixing updown.txt](https://github.com/gitbls/pistrong/files/13624542/Server.side.syslog.after.fixing.updown.txt)
[Client side syslog after fixing updown.txt](https://github.com/gitbls/pistrong/files/13624543/Client.side.syslog.after.fixing.updown.txt)

**Comment by gitbls** [comment]
Wow, I've never seen that before. Let's do a bit of diagnosis before you upgrade, as the upgrade will be a bit of work.

On both systems, sudo edit the file /etc/strongswan.d/charon-systemd.conf  and make the following edits:
* uncomment default=1
* comment out the lines enc=0, ike=0,net=0
* Save and exit

Now reboot both systems and make the VPN connection again. Grab the logs, and also run pscollect (from this github) on both systems. Send me all the info. Please email them to me (email can be found on https://github.com/gitbls/sdm; search for outlook.com on that page). Don't worry about redacting anything. I have way too much going on to spend time hacking your server 🤣

Thx

**Reply by rfraser578** [reply]
Thank you for that. I will get that done and tested as soon as I can and get back to you. It might be a day or two, because it's probably better I take the client Raspberry Pi outside of my own network for a real world test as the other connections have been done.
Kind regards,
Russell

**Comment by gitbls** [comment]
Ah, so you're currently set up with both systems inside your network? That's fine, but if the client is configured to connect to your public IP, that would require what is called "hair-pinning", where the client connection goes out your router and then back in through the router to the server.

Depending on the router, sometimes that works, and sometimes that doesn't.

If that's the case, what is the result if you change the `remote_addrs` in the CLIENT's /etc/swanctl/conf.d/Tunnel-xxxx.conf, and restart strongSwan on the client?

**Reply by rfraser578** [reply]
I am setup with both client and server machines inside my own network just now, but the other tests I have done have been with the client at another location.

I just did the /etc/strongswan.d/charon-systemd.conf edits and rebooted both Raspberry Pi's before your last message came in. Should I go ahead and change the remote_addrs and test again?

**Reply by gitbls** [reply]
If you've already got it set up, it's a really simple test to run and would provide additional information.

**Comment by rfraser578** [comment]
I changed the remote address to the network IP and launched the tunnel again, but the same result. I've emailed the syslogs and pscollect files.

**Comment by rfraser578** [comment]
Hi Benn,

Thanks for the diagnosis and direction in our email exchange. I worked out that I needed to use the iptables-persistent package which saves the data so that on reboot the iptables are reloaded. After I got that straightened, the tunnel worked.

Kind regards,

Russell

**Reply by gitbls** [reply]
Excellent! I have a few more things to add to `pscollect` based on what we worked out. Great working with you on this!
